### ISSUE TITLE
`tasks` broken?

### ISSUE THREAD

**Issue by vext01** [jjjake/internetarchive#167]
Hi,

I've configured internetarchive (I am able to start uploads), but `tasks` does not seem to work:

```
$ ia upload softdev_warmup_experiment_artefacts octane_spidermonkey_results_0_7_linux3_e3_1240.json.bz2.asc
softdev_warmup_experiment_artefacts:
 uploading octane_spidermonkey_results_0_7_linux3_e3_1240.json.bz2.asc: [################################] 1/1 - 00:00:00
$ ia tasks
error: unable to parse JSON. have you run `ia configure`?
```

This is OpenBSD-current with Python-2.7.12.

**Comment by waynedpj** [comment]
ahoy,

  seeing the same issue with a new install of v1.1.0 on Gentoo x86-64 GNU+Linux:

```
$ python --version
Python 3.4.3

$ ia tasks in-giro--demonstrate
error: unable to parse JSON. have you run `ia configure`?
```

but `ia configure` has been run successfully and other commands are working fine.

thanks, very handy utility.

**Comment by saper** [comment]
It seems to me that this is a server side issue.

https://archive.org/details/MarianRejewskiGenialnyKryptolog?output=json&json=2 seems to provide valid json, but

https://archive.org/history/MarianRejewskiGenialnyKryptolog?output=json&json=2

does not (it produces the same HTML without CGI parameters)

Similar with https://archive.org/catalog.php?json=2&output=json&callback=foo which seems to return the same HTML content as https://archive.org/catalog.php

Examples from the [archive.org JSON API spec](https://archive.org/help/json.php) do seem to work fine, is `ia` using some unofficial and unsupported endpoints? I think the idea behind that JSON API was to be able to ask for JSON for almost any URL.

**Comment by jjjake** [comment]
Getting the archive.org catalog to return JSON is a bit quirky.

Can you all confirm that this wasn't a transient error? I'm unable to reproduce this at the moment, but I have seen this error before and it was transient for me.

To get the catalog to return JSON, three params are required: `json=2`, `output=json`, and `callback=<foo>`.

@saper is https://archive.org/catalog.php?json=2&output=json&callback=foo still returning HTML for you? It's returning JSON wrapped in a callback for me at the moment.

**Comment by waynedpj** [comment]
@jjjake just tried it again and same error, whether i put in an identifier or not.  thanks for looking into this.

**Comment by saper** [comment]
I can confirm this is well:

both

https://archive.org/history/MarianRejewskiGenialnyKryptolog?output=json&json=2&callback=x

and

https://archive.org/catalog.php?json=2&output=json&callback=foo

seem to return text/html for me now

**Comment by jeffgerhard** [comment]
I haven't been able to get `tasks` to work for several days now. It seems to be a recurrence of the above problems. I can construct URLs that return JSON, i.e., https://archive.org/catalog.php?json=2&output=json&callback=foo seems fine.  But `ia tasks` and the python `get_tasks`  command fail with "unable to parse JSON" errors.

If this is a transient problem it's persisted for a while. Wanted to chime in here before seeking workarounds. Thanks!

**Comment by vext01** [comment]
Seems like this is still an issue with the latest release.

**Comment by pgporada** [comment]
I'm seeing the same problem here.
```
$ pip list | grep internetarchive
internetarchive (1.7.6)
```

```
$ ia tasks SMLRE261
error: unable to parse JSON. have you run `ia configure`?

$ ia list SMLRE261
SMLR-E261.afpk
SMLR-E261.mp3
SMLR-E261.ogg
SMLR-E261.png
SMLR-E261_spectrogram.png
SMLRE261_archive.torrent
SMLRE261_files.xml
SMLRE261_meta.sqlite
SMLRE261_meta.xml
_HOUR_00_SMLR-E261.afpk
_HOUR_01_SMLR-E261.afpk
```

**Comment by jjjake** [comment]
Have you run `ia configure` @pgporada?

**Comment by pgporada** [comment]
@jjjake Yes, I had run that command during my initial setup. Sorry for the delay.

**Comment by jjjake** [comment]
I'm closing this issue for now, as I think this issue was/is transient. Please feel free to reopen it if you continue to have issues.

**Comment by JustAnotherArchivist** [comment]
I'm seeing this issue right now with version 1.7.7.

**Comment by JustAnotherArchivist** [comment]
Thanks @jjjake.

It seems to be consistent. At least it doesn't work now either. I can set up something to test it regularly.

I added the print call to `internetarchive.catalog.Catalog._get_tasks`. IA returns HTML instead of JSON(P):

    b'<!DOCTYPE html>\n<html lang="en">\n\n  <head>\n    <title>Catalog - Outstanding Tasks</title>\n\n
    [...]
    <div class="box" style="border:2px solid gray;width:95%;">\n          <h1>Help with this page</h1>\n    \n          This page shows you the tasks related to your item(s) that have\n          not completed yet.\n
    [...]

The parameters passed to `self.session.get` are:

    self.url = 'https://archive.org/catalog.php'
    self.params = {'json': 2, 'output': 'json', 'justme': 1, 'callback': 'foo'}
    self.request_kwargs = {}

The returned HTML also reflects those parameters in the sort links, by the way:

    <a href="/catalog.php?output=json&json=2&callback=foo&justme=1&sort=identifier&sortOrder=1">identifier</a>

**Comment by jjjake** [comment]
@JustAnotherArchivist I'm wondering if your cookies are expired? Try running `ia configure` again, and see if that works.

**Comment by JustAnotherArchivist** [comment]
@jjjake I really doubt that for two reasons: the returned HTML page has my username in it, and other things requiring authorisation (uploading and changing metadata) work fine.

I just tested it anyway, and the result is still the same.

**Comment by jjjake** [comment]
To be clear, uploading and changing metadata will continue to work once your cookies expire. These commands use IA-S3 keys. `ia tasks` depends on cookies, which expire.

Ok, thanks for trying.

I'm still unable to replicate this (with priv'd and non-priv'd accounts), but I'll continue to try! Thanks again for the report, and my apologies for the inconvenience.

**Comment by JustAnotherArchivist** [comment]
Oh, I see. Wasn't aware of that difference. But yeah, the returned HTML also contains a list of tasks (which in my case is currently empty), not just the login message you get when accessing `/catalog.php` without being logged in.

Let me know if I can do anything further to help you with debugging this.

**Comment by jjjake** [comment]
@JustAnotherArchivist Can you give this patch a try?:

https://github.com/jjjake/internetarchive/commit/9e0b1dd575f57bb0ef06f598aeb791e663f9b6e0

I think your issues might have been related to encoding. Also, either way, can you let me know what version of Python you're using?

**Comment by JustAnotherArchivist** [comment]
I don't see how that commit would help. `CatalogTask` isn't instantiated until after the HTTP request (i.e. also after the `print` call I had added), and I'm only using a plain `ia tasks` without `--json`.

I tested it anyway, and as suspected, I'm still getting the same behaviour.

I can reproduce this on three separate machines: Python 3.4.2 on Debian Jessie, Python 3.5.3 on Debian Stretch, and Python 3.6.3 (custom installation) on Debian Wheezy. The machines are completely independent of each other: different networks, different countries, two continents even. The outcome is identical on all machines, i.e. the server returns HTML instead of JSON, and the HTML content indicate that I'm logged in correctly. The only differences between the HTML across the three machines are a timestamp, the order of the parameters in the table headers (which seems to be random), and the internal server/timing information near the end of the markup. (NB, I only tested 9e0b1dd on the Jessie machine.)

I was also able to reproduce this with curl using the cookies written by `ia`:

    curl -v -b 'logged-in-user=<email>; logged-in-sig=<...>' 'https://archive.org/catalog.php?json=2&output=json&callback=foo&justme=1'

... returns the same HTML as `ia` gets. I tried adding an `Accept: application/json` header, but that didn't change anything either.

All of this leads me to think that this is an issue on the server side, e.g. `catalog.php` ignoring the `json` parameter. But it's weird that you're unable to reproduce it.

**Comment by jjjake** [comment]
Ok, thanks for testing again! When trying to replicate, I ran into this encoding issue (it's an issue when there are unicode characters in the task args). I thought that might be what you're running into.

The person who maintains this endpoint doesn't understand what could be wrong (other than expired cookies), and I'm still failing to replicate it (on multiple platforms, with multiple python versions, with multiple completely un-priv'd accounts. I'm also not able to replicate it in a browser or with curl. I've tested on non IA networks as well.).

Please let me know if you're relying on tasks for something specific. I might be able to help figure out a workaround (generally speaking, most users shouldn't have to worry about the task history). Feel free to email me if that makes more sense (my email is on my profile).

I think the long-term solution for this is to develop a more robust API for retrieving task info and history. I will try to push this forward.

Again, I apologize for the inconvenience. Thanks for taking the time to test all of this out, and report back with detailed information. I'll keep you posted if I find anything else out.

**Comment by JustAnotherArchivist** [comment]
> I'm still failing to replicate it

Odd. For the record, I'm seeing it in Firefox as well.

> Please let me know if you're relying on tasks for something specific.

I was hoping to use `ia tasks` to check from a terminal what the progress on deriving my newly uploaded items is. Don't worry about finding a workaround; I can always just check it from the browser instead.

The main reason why I want to check this, by the way, is that I noticed that changing the `date` metadata field on an item (`ia metadata $identifier --modify='date:2018-05-01'`) while a derive is running doesn't work properly: once the derive finishes, that field is reset to the value it had before/when the derive was started. Other metadata, in particular the description (haven't tested other fields), is not affected by this. I've been told on ArchiveTeam IRC that this is a known issue and that it's not recommended to change metadata while derives are running. Hence why I, after uploading data to an item, want to check whether the derive is finished, i.e. whether I can safely change the metadata.

**Comment by jjjake** [comment]
@JustAnotherArchivist Do you have an example item where this happened? I'm not aware of this issue, and the person who maintains our Metadata API is not either.

The only case where I think this could happen is if you were deriving item metadata from a MARC record (and in that case, it would happen for any metadata field). Otherwise, this sounds like a bug that we're not aware of.

Feel free to email me, this conversation might be getting out of scope for this issue.

**Comment by saper** [comment]
I think I can reproduce this, too:

> env LANG=C ia-3.6 tasks
error: unable to parse JSON. have you run `ia configure`?

`ia list <someitem> works`

**Comment by JustAnotherArchivist** [comment]
@jjjake It appears that this error occurs when you don't have task admin access on IA. My account was given that recently, and now `ia tasks` works fine. dxrt confirmed in #internetarchive on EFnet that it's still failing with the reported error when the account doesn't have task admin access. I hope that helps you reproduce the bug.

**Comment by JustAnotherArchivist** [comment]
It looks like this issue has *finally* been fixed in v1.9.0 with the new tasks API!
